agent/taskresource/volume.go

```diff
@@ -0,0 +1,150 @@
+	Name                string
+	Mountpoint          string
+	Driver              string
+	DriverOptions       map[string]string
+	Labels              map[string]string
+	createdAt           time.Time
+	desiredStatusUnsafe VolumeStatus
+	knownStatusUnsafe   VolumeStatus
+	// lock is used for fields that are accessed and updated concurrently
+	lock sync.RWMutex
+}
+
+// NewVolumeResource returns a docker volume wrapper object
+func NewVolumeResource(name string, 
+	mountPoint string, 
+	driver string, 
+	driverOptions map[string]string, 
+	labels map[string]string) *VolumeResource {
+
+	return &VolumeResource{
+		Name: name,
+		Mountpoint: mountPoint,
+		Driver: driver,
+		DriverOptions: driverOptions,
+		Labels: labels,
+	}
+}
+
+// SetDesiredStatus safely sets the desired status of the resource
+func (vol *VolumeResource) SetDesiredStatus(status VolumeStatus) {
+	vol.lock.Lock()
+	defer vol.lock.Unlock()
+
+	vol.desiredStatusUnsafe = status
+}
+
+// GetDesiredStatus safely returns the desired status of the task
+func (vol *VolumeResource) GetDesiredStatus() VolumeStatus {
+	vol.lock.RLock()
+	defer vol.lock.RUnlock()
+
+	return vol.desiredStatusUnsafe
+}
+
+// SetKnownStatus safely sets the currently known status of the resource
+func (vol *VolumeResource) SetKnownStatus(status VolumeStatus) {
+	vol.lock.Lock()
+	defer vol.lock.Unlock()
+
+	vol.knownStatusUnsafe = status
+}
+
+// GetKnownStatus safely returns the currently known status of the task
+func (vol *VolumeResource) GetKnownStatus() VolumeStatus {
+	vol.lock.RLock()
+	defer vol.lock.RUnlock()
+
+	return vol.knownStatusUnsafe
+}
+
+// SetCreatedAt sets the timestamp for resource's creation time
+func (vol *VolumeResource) SetCreatedAt(createdAt time.Time) {
+	if createdAt.IsZero() {
+		return
+	}
+	vol.lock.Lock()
+	defer vol.lock.Unlock()
+
+	vol.createdAt = createdAt
+}
+
+// GetCreatedAt sets the timestamp for resource's creation time
+func (vol *VolumeResource) GetCreatedAt() time.Time {
+	vol.lock.RLock()
+	defer vol.lock.RUnlock()
+
+	return vol.createdAt
+}
+
+// volumeResourceJSON duplicates VolumeResource fields, only for marshalling and unmarshalling purposes
+type volumeResourceJSON struct {
+	Name          string            `json:"Name"`
+	Mountpoint    string            `json:"MountPoint"`
+	Driver        string            `json:"Driver"`
+	DriverOptions map[string]string `json:"DriverOptions"`
+	Labels        map[string]string `json:"Labels"`
+	DesiredStatus *VolumeStatus `json:"DesiredStatus"`
+	KnownStatus   *VolumeStatus `json:"KnownStatus"`
+}
+
+// MarshalJSON marshals VolumeResource object using duplicate struct VolumeResourceJSON
+func (vol *VolumeResource) MarshalJSON() ([]byte, error) {
+	if vol == nil {
+		return nil, nil
+	}
+	return json.Marshal(volumeResourceJSON{
+		vol.Name,
+		vol.Mountpoint,
+		vol.Driver,
+		vol.DriverOptions,
+		vol.Labels,
+		func() *VolumeStatus { desiredState := vol.GetDesiredStatus(); return &desiredState }(),
```

Comment:
can you check if either of `aws.Int(vol.GetDesiredStatus())` or `aws.Int32(vol.GetDesiredStatus())` work here? You should be able to reuse one of those

Reply:
It don't think I can, unless there's a way to cast *int to *VolumeStatus? 
Seems to generate as long a statement as this one.

Reply:
ah, never mind then.